src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,536 @@
+
+  auto adjust_lo = [](T lo, T zeros, T ones) {
+    constexpr size_t W = sizeof(T) * 8;
+    T zero_violation = lo & zeros;
+    T one_violation = ~lo & ones;
+    if (zero_violation == 0 && one_violation == 0) {
+      return lo;
+    }
+
+    if (zero_violation < one_violation) {
+      // Align the last violation of ones unset all the lower bits
+      // so we don't care about violations of zeros
+      juint last_violation = W - 1 - count_leading_zeros(one_violation);
+      T alignment = T(1) << last_violation;
+      lo = (lo & -alignment) + alignment;
+      return lo | ones;
+    }
+
+    // Suppose lo = 00110010, zeros = 01010010, ones = 10001000
+    // Since the 4-th bit must be 0, we need to align up the lower bound.
+    // This results in lo = 01000000, but then the 6-th bit does not match,
+    // align up again gives us 10000000.
+    // We can align up directly to 10000000 by finding the first place after
+    // the highest mismatch such that both the corresponding bits are unset.
+    // Since all bits lower than the alignment are unset we don't need to
+    // align for the violations of ones anymore.
+    juint last_violation = W - 1 - count_leading_zeros(zero_violation);
+    T find_mask = std::numeric_limits<T>::max() << last_violation;
+    T either = lo | zeros;
+    T tmp = ~either & find_mask;
+    T alignment = tmp & (-tmp);
+    lo = (lo & -alignment) + alignment;
+    return lo | ones;
+  };
+
+  T new_lo = adjust_lo(lo, zeros, ones);
+  if (new_lo < lo) {
+    empty = true;
+    return true;
+  }
+
+  T new_hi = ~adjust_lo(~hi, ones, zeros);
+  if (new_hi > hi) {
+    empty = true;
+    return true;
+  }
+  bool progress = (new_lo != lo) || (new_hi != hi);
+  lo = new_lo;
+  hi = new_hi;
+  empty = lo > hi;
+  return progress;
+}
+
+template <class T>
+static bool adjust_bits_from_bounds(bool& empty, T& zeros, T& ones, T lo, T hi) {
+  static_assert(std::is_unsigned<T>::value, "");
+  T mismatch = lo ^ hi;
+  T match_mask = mismatch == 0 ? std::numeric_limits<T>::max()
+                               : ~(std::numeric_limits<T>::max() >> count_leading_zeros(mismatch));
+  T new_zeros = zeros | (match_mask &~ lo);
+  T new_ones = ones | (match_mask & lo);
+  bool progress = (new_zeros != zeros) || (new_ones != ones);
+  zeros = new_zeros;
+  ones = new_ones;
+  empty = ((zeros & ones) != 0);
+  return progress;
+}
+
+template <class T>
+static void normalize_constraints_simple(bool& empty, T& lo, T& hi, T& zeros, T& ones) {
+  adjust_bits_from_bounds(empty, zeros, ones, lo, hi);
+  if (empty) {
+    return;
+  }
+  while (true) {
+    bool progress = adjust_bounds_from_bits(empty, lo, hi, zeros, ones);
+    if (!progress || empty) {
+      return;
+    }
+    progress = adjust_bits_from_bounds(empty, zeros, ones, lo, hi);
+    if (!progress || empty) {
+      return;
+    }
+  }
+}
+
+template <class T, class U>
+void normalize_constraints(bool& empty, T& lo, T& hi, U& ulo, U& uhi, U& zeros, U& ones) {
```

Comment:
With function signatures like this it is hard to know what are the arguments and what are the return values.